Modules/_copy.c

```diff
@@ -0,0 +1,432 @@
+            all_identical = 0;
+        PyTuple_SET_ITEM(y, i, copy);
+    }
+
+    /*
+     * Tuples whose elements are all their own deepcopies don't
+     * get memoized. Adding to the memo costs time and memory, and
+     * we assume that pathological cases like [(1, 2, 3, 4)]*10000
+     * are rare.
+     */
+    if (all_identical) {
+        Py_INCREF(x);
+        Py_SETREF(y, x);
+    }
+    /* Did we do a copy of the same tuple deeper down? */
+    else if ((z = _PyDict_GetItem_KnownHash(memo, id_x, hash_id_x)) != NULL)
+    {
+        Py_INCREF(z);
+        Py_SETREF(y, z);
+    }
+    /* Make sure any of our callers up the stack return this copy. */
+    else if (_PyDict_SetItem_KnownHash(memo, id_x, y, hash_id_x) < 0) {
+        Py_CLEAR(y);
+    }
+    return y;
+}
+
+
+#define ARRAY_SIZE(a) (sizeof(a)/sizeof((a)[0]))
+
+/*
+ * Using the private _PyNone_Type and _PyNotImplemented_Type avoids
+ * special-casing those in do_deepcopy.
+ */
+static PyTypeObject* const atomic_type[] = {
+    &_PyNone_Type,    /* type(None) */
+    &PyUnicode_Type,  /* str */
+    &PyBytes_Type,    /* bytes */
+    &PyLong_Type,     /* int */
+    &PyBool_Type,     /* bool */
+    &PyFloat_Type,    /* float */
+    &PyComplex_Type,  /* complex */
+    &PyType_Type,     /* type */
+    &PyEllipsis_Type, /* type(Ellipsis) */
+    &_PyNotImplemented_Type, /* type(NotImplemented) */
+};
+#define N_ATOMIC_TYPES ARRAY_SIZE(atomic_type)
+
+struct deepcopy_dispatcher {
+    PyTypeObject* type;
+    PyObject* (*handler)(PyObject* x, PyObject* memo,
+        PyObject* id_x, Py_hash_t hash_id_x);
+};
+
+static const struct deepcopy_dispatcher deepcopy_dispatch[] = {
+    {&PyList_Type, deepcopy_list},
+    {&PyDict_Type, deepcopy_dict},
+    {&PyTuple_Type, deepcopy_tuple},
+};
+#define N_DISPATCHERS ARRAY_SIZE(deepcopy_dispatch)
+
+static PyObject* do_deepcopy(PyObject* x, PyObject* memo)
+{
+    unsigned i;
+    PyObject* y, * id_x;
+    const struct deepcopy_dispatcher* dd;
+    Py_hash_t hash_id_x;
+
+    assert(PyDict_CheckExact(memo));
+
+    /*
+     * No need to have a separate dispatch function for this. Also, the
+     * array would have to be quite a lot larger before a smarter data
+     * structure is worthwhile.
+     */
+    for (i = 0; i < N_ATOMIC_TYPES; ++i) {
+        if (Py_TYPE(x) == atomic_type[i]) {
+            Py_INCREF(x);
+            return x;
+        }
+    }
+
+    /* Have we already done a deepcopy of x? */
+    id_x = object_id(x);
+    if (id_x == NULL)
+        return NULL;
+    hash_id_x = PyObject_Hash(id_x);
+
+    y = _PyDict_GetItem_KnownHash(memo, id_x, hash_id_x);
+    if (y != NULL) {
+        Py_DECREF(id_x);
+        Py_INCREF(y);
+        return y;
+    }
+    /*
+     * Hold on to id_x and its hash a little longer - the dispatch handlers
+     * will all need it.
+     */
+    for (i = 0; i < N_DISPATCHERS; ++i) {
+        dd = &deepcopy_dispatch[i];
+        if (Py_TYPE(x) != dd->type)
+            continue;
```

Comment:
It might be faster to replace the atomic types and dispatch types code with a ``_Py_hashtable_t`` table.

Reply:
``_hashopenssl.c`` has an example how to use the hash table API.

Reply:
@tiran Using valgrind I did profiling on various test cases. For none of the cases the lookup was a major bottleneck. The dispatch tables are small and private (so cannot grow). So unless we can find cases where the hash table approach would really help, I prefer to keep this implementation which is imo simpler than the hash table.